### ISSUE TITLE
[MPLUGIN-483] @Component in Mojo parent classes not detected/added to plugin.xml requirements

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-plugin-tools#789]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** opened **[MPLUGIN-483](https://issues.apache.org/jira/browse/MPLUGIN-483?redirect=false)** and commented

see https://github.com/dev-aspectj/aspectj-maven-plugin/pull/137#issuecomment-1757713422

---

**Affects:** 3.9.0

1 votes, 4 watchers

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Afaik Mojo annotations were never scanned in class hierarchy, and it was done on purpose.

Unlike jsr330 annotations, that are "inherited", Mojo annotations were never inherited.

**Comment by jira-importer** [comment]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** commented

[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas), I find it very hard to believe that anyone would do something like this on purpose, because it effectively means that developers have to redefine protected fields with those annotations, which in turn would lead to only the redefined ones being injected and the ones in the parent classes always null, rendering them utterly useless. It would be such a huge design flaw, that I am confident, nobody would do that on purpose. The whole concept of having a parent class like `AbstractMavenReport` with common fields and behaviours would fail.

BTW, of course annotations in the JVM cannot be inherited (except for ones on class types with the meta annotation `@Inherited`). It is not about inheriting them, rather about also scanning parent classes for their existence. They do not need to be inherited, because protected fields belong to (i.e. are visible and reachable via `this` from) subclasses.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

I'm also very surprised: we probably have cases where we have an abstract base Mojo class defining components and properties that are extended in many concrete Mojos

we need to check a few plugins with many goals around a few concepts: site? dependency?

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

We have also a ITs for such cases ... and looks working as expected:
https://github.com/apache/maven-plugin-tools/tree/master/maven-plugin-plugin/src/it

- annotation-with-inheritance-from-deps
- annotation-with-inheritance-reactor
- annotation-with-inheritance

**Comment by jira-importer** [comment]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** commented

Well, it is reproducibly not working in the project in the GitHub issue. See the findings there. The latest finding is: If I copy the source code of `org.apache.maven.reporting.AbstractMavenReport` into my project as class `org.apache.maven.reporting.AbstractMavenReportXXX` and make my `org.codehaus.mojo.aspectj.AjcReportMojo` extend `AbstractMavenReportXXX`, the integration test passes. If I switch back to extending class `AbstractMavenReport` from the external dependency, it fails again. Is there any mojo setting I need to use for this to work?

**Comment by jira-importer** [comment]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** commented

This issue can be closed. I found the root cause of the problem. A few years before I got involved in AspectJ Maven Plugin, someone added an inadequate workaround for MPLUGIN-328 to the code base, trying to address something completely different, namely an incompatible ASM version upstream in some Apache Maven plugin. The workaround looked as follows:

```xml
<plugin>
  <groupId>org.apache.maven.plugins</groupId>
  <artifactId>maven-plugin-plugin</artifactId>
  <version>${maven-plugin-plugin.version}</version>
  <configuration>
    <mojoDependencies>
      <mojoDependency>dev.aspectj:aspectj-maven-plugin</mojoDependency>
    </mojoDependencies>
  </configuration>
</plugin>
```

That old `mojoDependency` workaround was actually bad to begin with, because it only included `{}dev.aspectj:aspectj-maven-plugin{`}, therefore implicitly excluding other
dependencies like `maven-reporting-impl` where `AbstractMavenReport` lives. This in turn led to all super class `@Parameter` and `@Component` properties to not be included in the generated {_}plugin.xml{_}. I fixed this in the plugin build file, and now it works as expected. It also explains why my local copy of the base class inside the project worked: Because it was included in the scan.
